core/frontend/src/render/webgl/Texture.ts

```diff
@@ -480,6 +511,69 @@ export class Texture2DHandle extends TextureHandle {
   }
 }
 
+interface ExternalTextureRequest {
+  handle: Texture2DHandle;
+  name: string;
+  imodel: IModelConnection;
+  type: RenderTexture.Type;
+  format: ImageSourceFormat;
+}
+
+/** @internal */
+export class ExternalTextureLoader { /* currently exported for tests only */
+  public static readonly instance = new ExternalTextureLoader(10);
+  private readonly _maxActiveRequests: number;
+  private _numActiveRequests = 0;
+  private _pendingRequests: Array<ExternalTextureRequest> = [];
+
+  public get numActiveRequests() { return this._numActiveRequests; }
+  public get numPendingRequests() { return this._pendingRequests.length; }
+  public get maxActiveRequests() { return this._maxActiveRequests; }
+
+  private constructor(maxActiveRequests: number) {
+    this._maxActiveRequests = maxActiveRequests;
+  }
+
+  private _nextRequest() {
+    this._numActiveRequests--;
+    console.log("numActiveRequests = " + this._numActiveRequests);
+    if (this._numActiveRequests < this._maxActiveRequests && this._pendingRequests.length > 0) {
+      this._activateRequest(this._pendingRequests.pop()!);
+      console.log("pendingRequests.length = " + this._pendingRequests.length);
+    }
+  }
+
+  private _activateRequest(req: ExternalTextureRequest) {
+    this._numActiveRequests++;
+    console.log("numActiveRequests = " + this._numActiveRequests);
+    const texBytesPromise = req.imodel.getTextureImage({ name: req.name });
+    texBytesPromise.then((texBytes: Uint8Array | undefined) => {
+      if (undefined !== texBytes && !req.imodel.isClosed) {
+        const imageSource = new ImageSource(texBytes, req.format);
+        const imageElementPromise = imageElementFromImageSource(imageSource);
+        imageElementPromise.then((image: HTMLImageElement) => {
+          if (!req.imodel.isClosed) {
+            req.handle.reload(Texture2DCreateParams.createForImage(image, ImageSourceFormat.Png === req.format, req.type));
+            IModelApp.tileAdmin.invalidateAllScenes();
+          }
+          this._nextRequest();
+        }).catch((_e: any) => { });
+      } else
+        this._nextRequest();
+    }).catch((_e: any) => { });
+  }
+
+  public loadTexture(handle: Texture2DHandle, name: string, imodel: IModelConnection, type: RenderTexture.Type, format: ImageSourceFormat) {
+    console.log("loadTexture: " + name);

```

Comment:
Avoid having multiple requests for the same texture Id.

Reply:
Pushed changes that do this.